pkg/controller/nodeipam/ipam/cidrset/metrics.go

```diff
@@ -0,0 +1,89 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cidrset
+
+import (
+	"sync"
+
+	"k8s.io/component-base/metrics"
+	"k8s.io/component-base/metrics/legacyregistry"
+)
+
+const rangeAllocatorSubsystem = "range_allocator"
+
+var (
+	cidrSetAllocations = metrics.NewCounterVec(
+		&metrics.CounterOpts{
+			Subsystem:      rangeAllocatorSubsystem,
+			Name:           "cidrset_allocation_total",
```

Comment:
Because this is part of the mega controller-manager, maybe we should prefix the metric names with "nodeipam_" ?

Reply:
I was thinking on how to structure this correctly, let's see, because there are several dimensions. 

It seems we should keep node_ipam_controller as the subsystem, it has already registered here

https://github.com/kubernetes/kubernetes/blob/ded6ee953c68f8333ee6291e0bcb7e58604fac00/pkg/controller/nodeipam/node_ipam_controller.go#L101

then, the ipam controller may have a different allocator type

https://github.com/kubernetes/kubernetes/blob/dde6e8e7465468c32642659cb708a5cc922add64/pkg/controller/nodeipam/ipam/cidr_allocator.go#L41-L52

 and the cidrSet seems to be only used by the RangeAllocator, but each clusterCIDR (dual stack) will have its cidrset (we can add a new dimention using the label clusterCIDR)

I see 2 possibilities, because ... we may want to add metrics for the other components too?, counts the number of allocations in the rangeallocator and in the cidrset?

* adding a new dimension based on the node_ipam subcomponents and removing the cidset_prefix
```go
Subsystem: "node_ipam_controller",
Name:           "cidrs_allocations_total",
Label: "clusterCIDR", "subcomponent"
```
and subcomponent may be "cidrset", "rangeallocator", "cloudallocator"

* or prefix the metric with the subcomponent, we know that cidrsets are unique to rangeallocator
```go
Subsystem: "node_ipam_controller",
Name:           "cidrsets_cidrs_allocations_total",
Label: "clusterCIDR"
```

```go
Subsystem: "node_ipam_controller",
Name:           "rangeallocator_cidrs_allocations_total",
Label: "clusterCIDR"
```

First option sounds more flexible, but I think that it should require implement it at once to provide consistency along the codebase.
I'd go with the second one by now :)
(feel free to suggest better naming)

Reply:
I don't know best-practice here.

@brancz do you have guidance?